File: LeetCode/Medium/0053-maximum-subarray/0053-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        // dp[i] = 以 nums[i] 结尾的最大子数组和（局部的，有约束的）
        //globalMax = 整个数组中所有子数组的最大和（全局的，最终答案）
        int n = nums.length;
        int[] dp = new int[n];
        dp[0] = nums[0];
        int globalMax = dp[0];
        for (int i = 1; i < n; i++) {
            dp[i] = Math.max(dp[i - 1] + nums[i], nums[i]);
            globalMax = Math.max(globalMax, dp[i]);

        }
        return globalMax;
    }
}

File: LeetCode/Medium/0053-maximum-subarray/0053-maximum-subarray.py


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)

        dp_max = [0] * n
        dp_max[0] = nums[0]

        for i in range(1, n):
            dp_max[i] = max(nums[i], dp_max[i -1] + nums[i])

        return max(dp_max)

        